### PR TITLE
Done command

### LINKED ISSUE #27: Mark appointment as complete
As a nurse, I can mark the appointment as complete if the patient arrives for his/her appointment

### PR DESCRIPTION
**Pull request by YangYue128-helen**
Mark an appointment as done.

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/DoneCommand.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import seedu.address.model.appointment.Time;
+
+/**
+ * Marks an appointment in the list as done.
+ */
+public class DoneCommand extends Command {
+    public static final String COMMAND_WORD = "done";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Mark the appointment as done "
+            + "by date and time specified."
```
Comment (JinHao-L) [inline src/main/java/seedu/address/logic/commands/DoneCommand.java]: Maybe can rephrase this to `Mark the specified appointment as done` or `Mark the appointment specified by the date and time as done` (more specific version).
Lemme know what you think! 😄

File: src/main/java/seedu/address/logic/commands/DoneCommand.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+     */
+    private static Appointment createAppointment(AssignCommand.DateTimeLoader dateTimeLoader) {
+        assert dateTimeLoader.getDate().isPresent();
+        assert dateTimeLoader.getTime().isPresent();
+
+        Date assignedDate = dateTimeLoader.getDate().get();
+        Time assignedTime = dateTimeLoader.getTime().get();
+
+        return new Appointment(assignedDate, assignedTime);
+    }
```
Comment (JinHao-L) [inline src/main/java/seedu/address/logic/commands/DoneCommand.java]: I am not really sure if we should create an appointment without a patient for checking purpose. I made a PR to remove the optional field of Patient so an appointment cannot be made without a Patient (see #90) 

Maybe for the `filter` in line 60, can change it to check if an appointment has the same date and time? (`hasDate` and `hasTime`) Do you think this works?

File: src/main/java/seedu/address/logic/commands/DoneCommand.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    private final AssignCommand.DateTimeLoader dateTimeLoader;
+    /**
+     * Creates a DoneCommand to add a new {@code Appointment}
+     * @param dateTimeLoader details of an appointment.
+     */
+    public DoneCommand(AssignCommand.DateTimeLoader dateTimeLoader) {
+        requireAllNonNull(dateTimeLoader);
+        this.dateTimeLoader = new AssignCommand.DateTimeLoader(dateTimeLoader);
+    }
```
Comment (JinHao-L) [inline src/main/java/seedu/address/logic/commands/DoneCommand.java]: It seems wrong to use a nested class in AssignCommand for non-assign related command. 
Should we abstract out the DateTimeLoader into a class on its own? Maybe this can be addressed in future PRs.

File: src/main/java/seedu/address/model/AppointmentBook.java [hunk 4]
```diff
@@ -66,14 +66,18 @@ public void addAppointment(Appointment a) {
...
-    //  * The identity of {@code editedPatient} must not be the same as another existing appointment
-    //  * in the address book.
-    //  */
-    // public void setAppointment(Appointment target, Appointment editedAppointment) { }
-    //
+    /**
+     * Replaces the given appointment {@code target} in the list with {@code editedAppointment}.
+     * {@code target} must exist in the appointment book.
+     * The identity of {@code editedPatient} must not be the same as another existing appointment
+     * in the address book.
```
Comment (JinHao-L) [inline src/main/java/seedu/address/model/AppointmentBook.java]: `editedPatient` should be `editedAppointment` instead.

File: src/test/java/seedu/address/logic/commands/AddCommandTest.java [hunk 5]
```diff
@@ -159,6 +164,11 @@ public void setPatient(Patient target, Patient editedPatient) {
             throw new AssertionError("This method should not be called.");
         }
 
+        @Override
+        public void setAppointment(Appointment target, Appointment editedPatient) {
+
+        }
+
```
Comment (JinHao-L) [inline src/test/java/seedu/address/logic/commands/AddCommandTest.java]: For this can just add `throw new AssertionError("This method should not be called.");` similar to above

File: src/main/java/seedu/address/model/appointment/UniqueAppointmentList.java [hunk 6]
```diff
@@ -47,15 +56,25 @@ public void add(Appointment toAdd) {
...
+
+        int index = internalList.indexOf(target);
+        if (index == -1) {
+            throw new AppointmentNotFoundException();
+        }
+        if (target.equals(editedAppointment) && hasOverlaps(editedAppointment)) {
+            throw new OverlappingAppointmentException();
+        } else {
+            internalList.set(index, editedAppointment);
+        }
```
Comment (JinHao-L) [inline src/main/java/seedu/address/model/appointment/UniqueAppointmentList.java]: The check should probably be `!target.startAtSameTime(editedAppointment) && hasOverlaps(editedAppointment)` instead. (If the edited version is different from the target, make sure that there is no overlap with the rest of the appointment.)

I think that the `else {` part can also be removed so that the happy path is made more prominent. 😄

**Review by JinHao-L — COMMENTED** [review]
Other than the minor edits in the comments. LGTM 👍

File: src/main/java/seedu/address/logic/commands/DoneCommand.java (hunk 1 above)
Comment (xz0127) [reply src/main/java/seedu/address/logic/commands/DoneCommand.java]: I agree with Jinhao! 🥇

File: src/main/java/seedu/address/logic/commands/DoneCommand.java (hunk 2 above)
Comment (xz0127) [reply src/main/java/seedu/address/logic/commands/DoneCommand.java]: I think Jinhao is making a good point here. Personally i feel a Patient might be necessary for an appointment instance. Would it be better if we include a patient here besides the dateTimeLoader?

File: src/test/java/seedu/address/logic/commands/AddCommandTest.java (hunk 5 above)
Comment (xz0127) [reply src/test/java/seedu/address/logic/commands/AddCommandTest.java]: I have included this part in my recent PR

**Review by xz0127 — COMMENTED** [review]
Other than the comments mentioned, the rest LGTM! 👍

File: src/main/java/seedu/address/logic/commands/DoneCommand.java (hunk 3 above)
Comment (WadeForward) [reply src/main/java/seedu/address/logic/commands/DoneCommand.java]: That is possible.
Although the initial purpose of the DateTimeLoader class is just improve the OOP for the AssignCommandParser to pass the date and time, we can make it into an inheritance or other implementation to broaden the usage when necessary

File: src/main/java/seedu/address/logic/commands/DoneCommand.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+            + PREFIX_DATE + "12-Dec-2021 "
+            + PREFIX_TIME + "4:00PM";
+
+    public static final String MESSAGE_MARK_DONE_SUCCESS = "Marked Appointment as done: %1$s";
+    public static final String APPOINTMENT_DOES_NOT_EXISTS = "There is no appointment at this time slot";
+    public static final String APPOINTMENT_HAS_BEEN_MARKED = "The appointment has been marked already!";
+    public static final String DATE_MISSING = "The date of appointment is missing";
+    public static final String TIME_MISSING = "The time of appointment is missing";
+
+    private final AssignCommand.DateTimeLoader dateTimeLoader;
```
Comment (WadeForward) [inline src/main/java/seedu/address/logic/commands/DoneCommand.java]: If you want to get rid of the `AssignCommand.`, I guess it would better to import the `DateTimeLoader` class so the code looks more precise.

File: src/main/java/seedu/address/logic/commands/DoneCommand.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        Appointment toMark = appointmentToMark.orElse(null);
+        if (toMark == null) {
+            throw new CommandException(APPOINTMENT_DOES_NOT_EXISTS);
+        }
+
+        if (toMark.getIsDoneStatus()) {
+            throw new CommandException(APPOINTMENT_HAS_BEEN_MARKED);
+        }
+        Appointment doneAppointment = toMark.markAsDone();
+
```
Comment (WadeForward) [inline src/main/java/seedu/address/logic/commands/DoneCommand.java:69]: Will the white space be too much?

File: src/main/java/seedu/address/logic/parser/DoneCommandParser.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+            throw new ParseException(DoneCommand.TIME_MISSING);
+        }
+
+        AssignCommand.DateTimeLoader dateTimeLoader = new AssignCommand.DateTimeLoader();
+        Date date = ParserUtil.parseDate(argMultimap.getValue(PREFIX_DATE).get());
+        Time time = ParserUtil.parseTime(argMultimap.getValue(PREFIX_TIME).get());
+        dateTimeLoader.setAppointmentDate(date);
+        dateTimeLoader.setAppointmentTime(time);
+
+        return new DoneCommand(dateTimeLoader);
```
Comment (WadeForward) [inline src/main/java/seedu/address/logic/parser/DoneCommandParser.java:44]: do we still follow the user guide? coz from there the done should be with the use of the AppointmentId, so maybe it would be better to use AppointmentId class instead of dateTimeLoader?

**Review by WadeForward — COMMENTED** [review]
LGTM!

File: src/main/java/seedu/address/logic/commands/DoneCommand.java (hunk 7 above)
Comment (Avalionnet) [reply src/main/java/seedu/address/logic/commands/DoneCommand.java]: Yeah I agree with @ZhangWanlin98 that it may be better to import the `DateTimeLoader` class directly to improve code readability and keep the variables used on the same level of abstraction.

File: src/main/java/seedu/address/logic/commands/DoneCommand.java (hunk 2 above)
Comment (YangYue128-helen) [reply src/main/java/seedu/address/logic/commands/DoneCommand.java]: I changed the startAtSameTime method already. Check whether it looks good to you now.

File: src/main/java/seedu/address/logic/parser/DoneCommandParser.java (hunk 9 above)
Comment (YangYue128-helen) [reply src/main/java/seedu/address/logic/parser/DoneCommandParser.java:44]: I've checked with Jin Hao. The input of done command follows d/.. t/.. Later we can update the UG.

File: src/main/java/seedu/address/logic/commands/DoneCommand.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    public static final String MESSAGE_MARK_DONE_SUCCESS = "Marked Appointment as done: %1$s";
+    public static final String APPOINTMENT_DOES_NOT_EXISTS = "There is no appointment at this time slot";
+    public static final String APPOINTMENT_HAS_BEEN_MARKED = "The appointment has been marked already!";
+    public static final String DATE_MISSING = "The date of appointment is missing";
+    public static final String TIME_MISSING = "The time of appointment is missing";
+
+    private final DateTimeLoader dateTimeLoader;
+    /**
+     * Creates a DoneCommand to add a new {@code Appointment}
```
Comment (Avalionnet) [inline src/main/java/seedu/address/logic/commands/DoneCommand.java]: Perhaps it will be better to phrase it in the following way `Creates a DoneCommand to mark the specified appointment as done`? Let me know what you think of this :)

File: src/main/java/seedu/address/logic/parser/DoneCommandParser.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+import seedu.address.model.appointment.Date;
+import seedu.address.model.appointment.Time;
+
+/**
+ * Parses input arguments and creates a new DoneCommand object
+ */
+public class DoneCommandParser implements Parser<DoneCommand> {
+    /**
+     * Parses the given {@code String} of arguments in the context of the DoneCommand
+     * and returns an DoneCommand object for execution.
```
Comment (Avalionnet) [inline src/main/java/seedu/address/logic/parser/DoneCommandParser.java]: There seems to be some grammatical error here. It should be `returns a DoneCommand object` instead.

File: src/main/java/seedu/address/logic/parser/DoneCommandParser.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+     * @throws ParseException if the user input does not conform the expected format
```
Comment (Avalionnet) [inline src/main/java/seedu/address/logic/parser/DoneCommandParser.java]: Think you missed out a "to" and full stop here. Perhaps it will be better to phrase it as `if the user input does not conform to the expected format.`

**Review by Avalionnet — APPROVED** [review]
LGTM besides the minor comments raised. Good job 😄

**Review by Avalionnet — CHANGES_REQUESTED** [review]
Noticed some checks failing after my latest review. Do remember to fix them before merging :)

**Review by JinHao-L — APPROVED** [review]
LGTM! 👍

**Review by Avalionnet — APPROVED** [review]
LGTM 👍🏻
